From 424c4cac8c00b4e9a8a74eb2f1a56ebe95e48e6c Mon Sep 17 00:00:00 2001
From: Tsuchiya Yuto <kitakar@gmail.com>
Date: Sun, 28 Jun 2020 14:49:25 +0900
Subject: [PATCH 4/7] intel_soc_pmic_core: add parameter override_hrv

On Surface 3, when executing intel_soc_pmic_exec_mipi_pmic_seq_element(),
I recieve the following error:

    [ 7196.356682] intel_soc_pmic_exec_mipi_pmic_seq_element: Not implemented
    [ 7196.356686] intel_soc_pmic_exec_mipi_pmic_seq_element: i2c-addr: 0x6e reg-addr 0x57 value 0x63 mask 0xff

Surface 3 is an Atom Cherry Trail device and has PMIC device "INT33FD",
which is called "Crystal Cove PMIC". It seems that usually, _HRV of the
PMIC is 0x03 if device is Cherry Trail. However, DSDT of Surface 3 describes
the value is 0x02:

    Name (_HRV, 0x02)  // _HRV: Hardware Revision

due to this _HRV value, intel_pmic_bytcrc will be used as PMIC driver,
not intel_pmic_chtcrc.

It seems that the PMIC device "INT33FD" is located at i2c address 0x6e
on Surface 3. The latter driver defines pmic_i2c_address as 0x6e:

    static struct intel_pmic_opregion_data intel_chtcrc_pmic_opregion_data = {
        .pmic_i2c_address = 0x6e,
    };

but the former driver doesn't define the address, resulting in function
intel_soc_pmic_exec_mipi_pmic_seq_element() not working with errors
mentioned above.

If the latter driver is used, the functioin seems to work on Surface 3,
too.

This commit provides a way to override the _HRV value detected by
intel_soc_pmic_core driver.

TODO: Should Surface 3 really use the latter driver intel_pmic_chtcrc
      instead? or rather the pmic_i2c_address should be added to the
      former driver intel_pmic_bytcrc, too?

Usage:

    Add string like the following to your bootloader:

    intel_soc_pmic.override_hrv=Y intel_soc_pmic.override_hrv_value=0x03

Reference: cc0594c4b0ef ("ACPI / PMIC: Add i2c address for thermal control")
Signed-off-by: Tsuchiya Yuto <kitakar@gmail.com>
---
 drivers/mfd/intel_soc_pmic_core.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/mfd/intel_soc_pmic_core.c b/drivers/mfd/intel_soc_pmic_core.c
index ddd64f9e3341..9b13c58531a4 100644
--- a/drivers/mfd/intel_soc_pmic_core.c
+++ b/drivers/mfd/intel_soc_pmic_core.c
@@ -28,6 +28,11 @@ static struct pwm_lookup crc_pwm_lookup[] = {
 	PWM_LOOKUP("crystal_cove_pwm", 0, "0000:00:02.0", "pwm_pmic_backlight", 0, PWM_POLARITY_NORMAL),
 };
 
+bool override_hrv;
+module_param(override_hrv, bool, 0400);
+int override_hrv_value;
+module_param(override_hrv_value, int, 0400);
+
 static int intel_soc_pmic_i2c_probe(struct i2c_client *i2c,
 				    const struct i2c_device_id *i2c_id)
 {
@@ -48,6 +53,9 @@ static int intel_soc_pmic_i2c_probe(struct i2c_client *i2c,
 		return -ENODEV;
 	}
 
+	if (override_hrv)
+		hrv = override_hrv_value;
+
 	switch (hrv) {
 	case BYT_CRC_HRV:
 		config = &intel_soc_pmic_config_byt_crc;
-- 
2.27.0

